eth_abi/grammar.py

```diff
@@ -291,22 +303,31 @@ def __init__(self, base, sub=None, arrlist=None, *, node=None):
         type.
         """
 
-    def to_type_str(self):
-        sub, arrlist = self.sub, self.arrlist
-
-        if isinstance(sub, int):
-            sub = str(sub)
-        elif isinstance(sub, tuple):
-            sub = "x".join(str(s) for s in sub)
-        else:
-            sub = ""
-
-        if isinstance(arrlist, tuple):
-            arrlist = "".join(repr(list(a)) for a in arrlist)
-        else:
-            arrlist = ""
+        self._type_str: Optional[str] = None
+        """
+        The string representation of an ABI type.  Once populated,
+        this will be equal to the type string from which it was created.
+        """
 
-        return self.base + sub + arrlist
+    def to_type_str(self):
+        type_str = self._type_str
+        if type_str is None:
+            sub, arrlist = self.sub, self.arrlist
+    
+            if isinstance(sub, int):
+                sub = str(sub)
+            elif isinstance(sub, tuple):
+                sub = "x".join(str(s) for s in sub)
+            else:
+                sub = ""
+    
+            if isinstance(arrlist, tuple):
+                arrlist = "".join(repr(list(a)) for a in arrlist)
+            else:
+                arrlist = ""
+    
+            type_str = self._type_str = self.base + sub + arrlist
+        return type_str
 
     @property
```

Comment:
is there a reason we don't use cached_property here?

Reply:
It won't let me comment in the correct place, so I'll ask here. 

secondly, is there a reason we don't use cached_property on is_dynamic (or set it during init)?

lastly, is there a reason we don't cache the result of the normalize function at the bottom?